In [79]:
# Source, attribution, credits
# https://www.udemy.com/deep-learning-prerequisites-the-numpy-stack-in-python/
# https://github.com/lazyprogrammer/machine_learning_examples/tree/master/numpy_class

In [37]:
import numpy as np
from datetime import datetime

## Lists vs Arrays

In [2]:
L = [1, 2, 3]
A = np.array([1, 2, 3])

In [6]:
L.append(4)

In [7]:
A.append(4)

AttributeError: 'numpy.ndarray' object has no attribute 'append'

In [4]:
L = L + [5]

In [5]:
A = A + [4, 5]

ValueError: operands could not be broadcast together with shapes (3,) (2,) 

In [8]:
L2 = []

In [9]:
for e in L:
    L2.append(e + e)

In [10]:
L2

[2, 4, 6, 8, 10, 8]

In [11]:
A + A

array([2, 4, 6])

In [12]:
2 * A

array([2, 4, 6])

In [13]:
2 * L

[1, 2, 3, 4, 5, 4, 1, 2, 3, 4, 5, 4]

In [14]:
L**2

TypeError: unsupported operand type(s) for ** or pow(): 'list' and 'int'

In [17]:
L2 = []
for e in L:
    L2.append(e*e)

In [18]:
L2

[1, 4, 9, 16, 25, 16]

In [19]:
A**2

array([1, 4, 9])

In [20]:
np.sqrt(A)

array([ 1.        ,  1.41421356,  1.73205081])

In [21]:
np.log(A)

array([ 0.        ,  0.69314718,  1.09861229])

In [22]:
np.exp(A)

array([  2.71828183,   7.3890561 ,  20.08553692])

**Avoid for loops when you can**

## Dot Product

In [24]:
a = np.array([1, 2])
b = np.array([2, 1])

In [27]:
dot = 0
for e, f in zip(a,b):
    dot += e*f
dot

4

In [28]:
a*b

array([2, 2])

In [29]:
np.sum(a*b)

4

In [30]:
(a*b).sum()

4

In [31]:
np.dot(a,b)

4

In [32]:
a.dot(b)

4

In [33]:
amag = np.sqrt((a*a).sum())
amag

2.2360679774997898

In [34]:
amag = np.linalg.norm(a)
amag

2.2360679774997898

In [35]:
cosangle = a.dot(b) / (np.linalg.norm(a) * np.linalg.norm(b))
cosangle

0.79999999999999982

In [36]:
angle = np.arccos(cosangle)
angle
# in radians by default

0.6435011087932847

### Comparing speed

In [38]:
# https://github.com/lazyprogrammer/machine_learning_examples/blob/master/numpy_class/python3/dot_for.py

a = np.random.randn(100)
b = np.random.randn(100)
T = 100000

def slow_dot_product(a, b):
    result = 0
    for e, f in zip(a, b):
        result += e*f
    return result

def fast_dot_product(a, b):
    return a.dot(b)

t0 = datetime.now()
for t in range(T):
    slow_dot_product(a, b)
dt1 = datetime.now() - t0

t0 = datetime.now()
for t in range(T):
    fast_dot_product(a, b)
dt2 = datetime.now() - t0

print("dt1 / dt2:", dt1.total_seconds() / dt2.total_seconds())

dt1 / dt2: 26.56195627132404


## Matrices and Vectors

In [39]:
M = np.array([[1,2], [3,4]])
L = [[1,2], [3,4]]

In [40]:
print(L[0])
print(L[0][0])

[1, 2]
1


In [41]:
print(M[0,0])

1


In [43]:
M2 = np.matrix([[1,2], [3,4]])
M2

matrix([[1, 2],
        [3, 4]])

In [45]:
A = np.array(M2)
A

array([[1, 2],
       [3, 4]])

In [46]:
A.T

array([[1, 3],
       [2, 4]])

**A matrix is really just a two dimensional numpy array. Apparently the numpy source insists on the usage of arrays over matrices (source?).**  
**So a matrix in numpy is really just a 2D array, and a vector is a 1D array. You can also say a matrix is a 2D vector.**

### Generating matrices to work with

In [47]:
Z = np.zeros(10)
Z

array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.])

In [49]:
Z = np.zeros((10, 10))
Z

array([[ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.]])

In [51]:
R = np.random.random((10, 10))
R
# Uniformly distributed

array([[ 0.21902595,  0.36148252,  0.05482029,  0.23013497,  0.58426261,
         0.78347615,  0.18611372,  0.03525433,  0.1057157 ,  0.06866156],
       [ 0.55380509,  0.84294429,  0.34459511,  0.48808731,  0.46529691,
         0.86876667,  0.61202168,  0.5997753 ,  0.57627795,  0.27082885],
       [ 0.38124781,  0.7245691 ,  0.46645072,  0.54635327,  0.80476675,
         0.92123521,  0.91083992,  0.99872648,  0.87091085,  0.60184653],
       [ 0.47464801,  0.06264738,  0.13339878,  0.1534557 ,  0.89447695,
         0.52210544,  0.7261583 ,  0.09674899,  0.35037264,  0.30874214],
       [ 0.57695315,  0.53815263,  0.00806794,  0.92942591,  0.43700439,
         0.3169959 ,  0.21885033,  0.72809408,  0.02576523,  0.90825495],
       [ 0.74626248,  0.52178003,  0.45009544,  0.44994706,  0.77999272,
         0.81885804,  0.48453883,  0.88133555,  0.91163316,  0.48097535],
       [ 0.27764075,  0.14666914,  0.84976066,  0.26570372,  0.31235514,
         0.07272876,  0.29946344,  0.63763817

In [52]:
G = np.random.randn((10, 10))

TypeError: 'tuple' object cannot be interpreted as an integer

In [71]:
G = np.random.randn(10, 12) # Notice the odd function signature
#  10 is the number of samples, 12 is the number of dimensions
G
# Mean 0 Unit variance gaussian

array([[ 0.04990459, -0.32950169, -0.93206652,  0.7702643 ,  0.19096985,
         1.43717708, -0.15930853, -0.42596672,  1.10028873,  1.28441038,
        -1.13043114,  0.4198426 ],
       [-1.37093373, -2.01395387, -1.26779553, -1.16220865,  1.04998192,
         0.1394443 , -0.38564884,  1.26306055,  0.40536353, -0.61690283,
         1.65676804,  0.47624729],
       [ 1.84617267, -1.87812464, -1.60749622,  0.95852782,  0.84209969,
         1.30830137, -0.8105951 , -0.22322449, -0.496316  , -0.45496832,
         0.20848033,  0.55085717],
       [-0.2483376 ,  0.94577246,  0.52715526, -0.47382876,  0.45948801,
         1.2200367 , -0.25857355, -1.02407911,  1.25271976, -1.67907843,
         0.71629387, -0.93065768],
       [ 0.3035355 ,  0.36993114,  0.33910193, -0.43183919,  0.6183417 ,
        -0.0607555 , -0.06866072, -0.0589463 ,  2.36420614, -0.34822985,
        -0.98593408,  2.33510329],
       [ 1.0461655 ,  1.03529625,  0.00792943,  1.81802442,  0.30112279,
         0.15385682,  

In [56]:
print(G.mean())
print(G.var())

0.0102837626607
1.04286120978


### Matrix operations

#### Matrix Products
Use np.dot for inner product. Inner dimension should match, of course.

Keep in mind that * does elementwise multiplication when you're using arrays.

**Outer product**:
$C(i,j) = A(i)B(j)$

In [64]:
a = np.array([1, 2])
b = np.array([3, 4])
np.outer(a, b)

array([[3, 4],
       [6, 8]])

#### Matrix Inverse

In [58]:
A = np.array([[1,2], [3,4]])
Ainv = np.linalg.inv(A)
Ainv

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

In [59]:
Ainv.dot(A)

array([[  1.00000000e+00,   0.00000000e+00],
       [  2.22044605e-16,   1.00000000e+00]])

In [60]:
A.dot(Ainv)

array([[  1.00000000e+00,   0.00000000e+00],
       [  8.88178420e-16,   1.00000000e+00]])

#### Matrix determinant

In [61]:
np.linalg.det(A)

-2.0000000000000004

#### Matrix diagonal

In [62]:
np.diag(A)

array([1, 4])

In [63]:
np.diag([1,2])

array([[1, 0],
       [0, 2]])

**If you pass a 2D array to np.diag, you get a 1D array representing the diagonal elements. If you pass a 1D array, you get a 2D array with the all elements zero except diagonal elements equal to the 1D array.**

#### Trace of a matrix

In [65]:
np.trace(A)

5

#### Eigenvalues and Eigenvectors

In [66]:
X = np.random.randn(100, 3)

In [67]:
cov = np.cov(X)

In [68]:
cov.shape

(100, 100)

In [72]:
cov = np.cov(X.T)
cov.shape

(3, 3)

**To find the covariance of a data matrix, you want to transpose it first**

eigenvalues, eigenvectors = np.eig(C)  

OR  

eigenvalues, eigenvectors = np.eigh(C)  
for symmetric ($A = A^T$) and hermitian ($A = A^H$) matrices, $A^H = $ conjugate transpose of A.

In [73]:
np.linalg.eigh(cov)

(array([ 0.66352528,  0.76238285,  1.18517814]),
 array([[ 0.59712327,  0.07000267, -0.79908912],
        [ 0.7912459 ,  0.1122914 ,  0.60109946],
        [-0.1318094 ,  0.99120647, -0.01166249]]))

In [74]:
np.linalg.eig(cov)

(array([ 1.18517814,  0.66352528,  0.76238285]),
 array([[ 0.79908912, -0.59712327,  0.07000267],
        [-0.60109946, -0.7912459 ,  0.1122914 ],
        [ 0.01166249,  0.1318094 ,  0.99120647]]))

### Solving a linear system

Problem: $Ax = b$  
Solution: $A^{-1}Ax = x = A^{-1}b$  
is a system of D equations and D unknowns, if A is D x D.

Assume that A is invertible.

In [75]:
A = np.array([[1,2],[3,4]])

In [76]:
b

array([3, 4])

In [77]:
x = np.linalg.inv(A).dot(b)
x

array([-2. ,  2.5])

In [78]:
x = np.linalg.solve(A, b)
x

array([-2. ,  2.5])

**Always use solve, and not the inverse method**

The admission fee at a small fair is \$1.50 for children and \$4.00 for adults. On a certain day, 2200 people enter the fair and \$5050 is collected. How many children and how many adults attended.

Let:  
X1 = number of children, X2 = number of adults  

X1 + X2 = 2200  
1.5X1 + 4X2 = 5050

In [81]:
A = np.array([[1,1], [1.5,4]])
b = np.array([2200, 5050])
np.linalg.solve(A, b)

array([ 1500.,   700.])